### ISSUE TITLE
Autograph applied to Keras Custom Loss during Eager Execution

### ISSUE THREAD

**Issue by MrManiaLive** [tensorflow/tensorflow#41189]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10, x64
- TensorFlow installed from (source or binary): binary (pip)
- TensorFlow version (use command below): TF 2.4.0.dev20200707
- Python version: 3.7
- CUDA/cuDNN version: 10.1 / 7.6
- GPU model and memory: Bug appears on several computers with different GPU

**Describe the current behavior**
Tensorflow applies AutoGraph to keras custom loss even in eager execution, meaning that we can't debug the loss anymore (unless using tf.print). This did not happen in previous versions of Tensorflow.
Notice that it both happens when run_eagerly is set to True in model.compile() and when tf.config.run_functions_eagerly is set to True.

**Describe the expected behavior**
When run_eagerly=True is passed to the model during compilation, we should expect Tensorflow to run eagerly in the loss function.

**Standalone code to reproduce the issue**

```
import numpy as np
import tensorflow as tf
from tensorflow import keras

# Custom Model. Autograph is not applied in eager execution so debugging is possible.
class CustomModel(keras.models.Model):
	def __init__(self):
		super(CustomModel, self).__init__()
		self.layer = tf.keras.layers.Dense(3) # Can debug here

	def call(self, inputs, training=None, mask=None):
		x = self.layer(inputs) # Can debug here
		return x

# Custom Loss. AutoGraph is applied in eager execution so debugging is impossible.
class CustomLoss(keras.losses.Loss):
	def call(self, y_true, y_pred):
		x = tf.reduce_mean(tf.abs(y_pred-y_true)) # Cannot debug here
		return x

if __name__ == '__main__':
	data = np.random.random((1000, 3)).astype(np.float32)

	model = CustomModel()

	model.compile(loss=CustomLoss(), run_eagerly=True)
	model.fit(x=data, y=data, batch_size=32)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@MrManiaLive 
I ran the code shared, loss in [tf 2.2 is as per this gist](https://colab.research.google.com/gist/Saduf2019/fe531dd75ea910a954ac8525a6f2073e/untitled267.ipynb), and loss in [tf_nightly is as per this gist](https://colab.research.google.com/gist/Saduf2019/9dcccd4deaaf83c5e03076e16a9bff79/untitled268.ipynb).
Please let us know if this confirms your issue.

**Comment by nikitamaia** [comment]
@MrManiaLive can you also explain how you can tell that it's not running eagerly? With tf nightly I am able to print `tf.executing_eagerly()` within the loss function, which returns as `True`. Some more information on what debugging you are not able to do now would be useful.

**Comment by MrManiaLive** [comment]
> @MrManiaLive can you also explain how you can tell that it's not running eagerly? With tf nightly I am able to print `tf.executing_eagerly()` within the loss function, which returns as `True`. Some more information on what debugging you are not able to do now would be useful.

Within Pycharm, when I put a breakpoint inside the call method of the custom loss function, the program does not stop (whereas it does in the custom model). I though this meant that the loss function was autographed. I am left with tf.print() for debugging.

Notice that I am using the latest version of tf-nightly-gpu (tf 2.4). I remember not having this issue with tf 2.2.

**Comment by nikitamaia** [comment]
I'm not very familiar with Pycharm so I cannot speak to that, but I don't see any signs that the loss function is not running eagerly. I can also print `x.numpy` within the loss function and that works fine when `run_eagerly=True`

**Comment by MrManiaLive** [comment]
You're right. I tried to print x.numpy in the call method and it worked. I also printed tf.executing_eagerly() and it returned True. However, I still can't stop in the call method within Pycharm.

I am investigating the code.
In the `__call__` method of the parent class `Loss` `autograph.tf_convert()` is called.

And I have `ctx.status` equal to `ag_ctx.Status.UNSPECIFIED`, which creates a wrapper_factory.
It then calls `converted_call `in the `api.py` file and runs these lines:

  ```
try:
    program_ctx = converter.ProgramContext(
        options=options, autograph_module=tf_inspect.getmodule(converted_call))
    converted_f = conversion.convert(target_entity, program_ctx)
    if logging.has_verbosity(2):
      _log_callargs(converted_f, effective_args, kwargs)
  except Exception as e:  # pylint:disable=broad-except
    logging.log(1, 'Error transforming entity %s', target_entity, exc_info=True)
    if is_autograph_strict_conversion_mode():
      raise
    return _fall_back_unconverted(f, args, kwargs, options, e)

  with StackTraceMapper(converted_f), tf_stack.CurrentModuleFilter():
    try:
      if kwargs is not None:
        result = converted_f(*effective_args, **kwargs)
      else:
        result = converted_f(*effective_args)
    except Exception as e:
      _attach_metadata(e, converted_f)
      raise
```

While `target_entity `is my `CustomLoss.call`, `converted_f `is a `<function outer_factory.<locals>.inner_factory.<locals>.tf__call at 0x000001F5A2D6B558>`

Do you think this could be the source of my problem?

**Comment by nikitamaia** [comment]
Hi @MrManiaLive, sorry for the delay. I'm not an expert with Pycharm, you might want to submit a request on [official Pycharm forum](https://youtrack.jetbrains.com/issues/PY?_ga=2.9618933.1109719345.1596474586-855270654.1596474586)
If you are still unable to figure it out at that point, please update this thread and we can investigate if this is an issue on the TF side.

**Comment by jimthompson5802** [comment]
I'm encountering the same issue.  PyCharm breakpoints are recognized in TF2.2 but some breakpoints are not recognized in TF2.3.  Here is my minimal working example of the problem.   Although I have not explicitly tested different Python IDEs (e.g., VSCode), I believe this problem will occur in those other IDE with interactive debuggers.
```
import tensorflow as tf

tf.config.experimental_run_functions_eagerly(True)


class MyLoss(tf.keras.losses.Loss):
    def __init__(self):
        super().__init__()

    def call(self, y_true, y_pred):
        mae = tf.abs(y_true - y_pred)     # set breakpoint here, debugger stops for tf2.2 only
                                          # breakpoint not recognized in tf2.3
        return mae


@tf.function
def test_debug_breakpoint(x, y):
    mae = tf.abs(x - y)        # set breakpoint here, debugger stops for both tf2.2 and tf2.3
    return tf.reduce_mean(mae)


print("tf version:", tf.__version__)

t1 = tf.random.normal([10, 5], dtype=tf.float32)
t2 = tf.random.normal([10, 5], dtype=tf.float32)

# test user function
ans = test_debug_breakpoint(t1, t2)
print('test_debug_breakpoint:',ans.numpy())

# test custom Loss function
my_loss = MyLoss()
ans = my_loss(t1, t2)
print('custom loss function:', ans.numpy())

```

With the help of a colleague found reference to this experimental function: [`tf.autograph.experimental.do_not_convert()`](https://www.tensorflow.org/api_docs/python/tf/autograph/experimental/do_not_convert).  By using this  decorator as shown below, the unrecognized breakpoint is now recognized by the PyCharm debugger.
```
class MyLoss(tf.keras.losses.Loss):
    def __init__(self):
        super().__init__()

    @tf.autograph.experimental.do_not_convert   # breakpoint now recognized with this decorator
    def call(self, y_true, y_pred):
        mae = tf.abs(y_true - y_pred)     # set breakpoint here, debugger stops for tf2.2
                                          # debugger stops at this breakpoint in tf2.3
        return mae
```

In TF 2.2, it was possible to recognize  breakpoints, regardless of location, with only the use of this function: [`tf.config.experimental_run_functions_eagerly(True)`](https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/config/experimental_run_functions_eagerly)

In TF 2.3, it appears I need to know use both `tf.config.run_functions_eagerly(True)` and the `tf.autograph.experimental.do_not_convert()` functions to allow PyCharm to recognize all breakpoints.  From my perspective having to use `tf.autograph.experimental.do_not_convert()` is redundant. 

* Is there documentation on how to use the two functions?
* Is there guidance on how we can use interactive debuggers with TF 2.3?

Any guidance will be appreciated.

**Comment by nikitamaia** [comment]
Hi @jimthompson5802 thanks for providing the code and also a workaround. I am reopening this issue for further investigation.

**Comment by mdanatg** [comment]
This looks like a bug. When calling `tf.config.run_functions_eagerly(True)`, AutoGraph should not be called at all. We can confirm this by calling `tf.autograph.set_verbosity(3, True)`, which results in this log entry:

```
INFO:tensorflow:Converted call: <bound method MyLoss.call of <__main__.MyLoss object at 0x7f6bfdead198>>
```

That log entry is unexpected.

@tomerk @omalleyt12 this looks like a regression, could you investigate the Keras calling site? The behavior of `call` should be consistent with that of `test_debug_breakpoint`.

@jimthompson5802 another workaround could be to decorate `call` with `@tf.function`, although that might have other unintended side effects.

**Comment by omalleyt12** [comment]
I have a PR I'm working on to fix this

**Comment by jimthompson5802** [comment]
Just checking in.  Any progress on fixing the issue?

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/daeec88c51ccca7e2f08880c612caa0d/untitled129.ipynb)..Thanks !

**Comment by nectario** [comment]
I am facing the same issue. Is this resolved?

**Comment by JesseRMeyer** [comment]
Seeing this issue on Visual Studio Code in WSL with TF 2.15.1.

**Comment by Kayyuri** [comment]
Hi,
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
